libafl_qemu/src/qemu/qemu_config.rs

```diff
@@ -0,0 +1,404 @@
+use core::{
+    fmt,
+    fmt::{Display, Formatter},
+};
+use std::{
+    path::{Path, PathBuf},
+    sync::OnceLock,
+};
+
+use getset::Getters;
+use libafl_derive;
+use strum_macros;
+use typed_builder::TypedBuilder;
+
+use crate::{Qemu, QemuInitError};
+
+pub(super) static QEMU_CONFIG: OnceLock<QemuConfig> = OnceLock::new();
+
+#[cfg(emulation_mode = "systemmode")]
+#[allow(non_camel_case_types)]
+#[derive(Debug, strum_macros::Display, Clone)]
+#[strum(prefix = "-accel ")]
+pub enum Accelerator {
+    kvm,
+    tcg,
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Debug, strum_macros::Display, Clone)]
+#[strum(prefix = "if=")]
+pub enum DriveInterface {
+    floppy,
+    ide,
+    mtd,
+    none,
+    pflash,
+    scsi,
+    sd,
+    virtio,
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Debug, strum_macros::Display, Clone)]
+#[strum(prefix = "format=")]
+pub enum DiskImageFileFormat {
+    qcow2,
+    raw,
+}
+
+#[derive(Debug, Clone, Default, TypedBuilder)]
+pub struct Drive {
+    #[builder(default, setter(strip_option, into))]
+    file: Option<PathBuf>,
+    #[builder(default, setter(strip_option))]
+    format: Option<DiskImageFileFormat>,
+    #[builder(default, setter(strip_option))]
+    interface: Option<DriveInterface>,
+}
+
+impl Display for Drive {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "-drive")?;
+
+        let mut is_first_option = true;
+        let mut separator = || {
+            if is_first_option {
+                is_first_option = false;
+                " "
+            } else {
+                ","
+            }
+        };
+
+        if let Some(file) = &self.file {
+            write!(f, "{}file={}", separator(), file.to_str().unwrap())?;
+        }
+        if let Some(format) = &self.format {
+            write!(f, "{}{format}", separator())?;
+        }
+        if let Some(interface) = &self.interface {
+            write!(f, "{}{interface}", separator())?;
+        }
+
+        Ok(())
+    }
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Debug, strum_macros::Display, Clone)]
+#[strum(prefix = "-serial ")]
+pub enum Serial {
+    none,
+    null,
+    stdio,
+}
+
+#[allow(non_camel_case_types)]
+#[derive(Debug, strum_macros::Display, Clone)]
+#[strum(prefix = "-monitor ")]
+pub enum Monitor {
```

Comment:
Monitor and Serial have exactly the same values, It'd make sense to combine them to `OutputKind` or similar.

Also, what's the difference of none and null?

Also, shouldn't be enums all caps?

Reply:
> Monitor and Serial have exactly the same values, It'd make sense to combine them to `OutputKind` or similar.

It's not clear to me how you would deduplicate it, can you elaborate more?

> Also, what's the difference of none and null?

They are 2 different settings in qemu, basically `none` means no device at all while `null` is discard its output. https://www.qemu.org/docs/master/system/invocation.html

Reply:
Yes with the prefix hack I guess you can not

Reply:
Ah you _could_ have a common enum and then implement `From` for the individual enums but it might be overcomplicated without clear benefit